zcash_primitives/src/note_encryption.rs

```diff
@@ -192,36 +222,41 @@ pub fn try_sapling_note_decryption(
     )
     .ok()?;
 
-    let mut d = [0u8; 11];
-    d.copy_from_slice(&plaintext[1..12]);
-
-    let v = (&plaintext[12..20]).read_u64::<LittleEndian>().ok()?;
-
-    let mut rcm = FsRepr::default();
-    rcm.read_le(&plaintext[20..52]).ok()?;
-    let rcm = Fs::from_repr(rcm).ok()?;
+    let (note, to) = parse_note_plaintext_minus_memo(ivk, cmu, &plaintext)?;
 
     let mut memo = [0u8; 512];
     memo.copy_from_slice(&plaintext[52..564]);
 
-    let diversifier = Diversifier(d);
-    let pk_d = match diversifier.g_d::<Bls12>(&JUBJUB) {
-        Some(g_d) => g_d.mul(ivk.into_repr(), &JUBJUB),
-        None => {
-            // Invalid diversifier in note plaintext
-            return None;
-        }
-    };
+    Some((note, to, Memo(memo)))
+}
 
-    let to = PaymentAddress { pk_d, diversifier };
-    let note = to.create_note(v, rcm, &JUBJUB).unwrap();
+/// Attempts to decrypt and validate the first 52 bytes of `enc_ciphertext` using the
+/// given `ivk`. If successful, the corresponding Sapling note is returned, along with the
+/// `PaymentAddress` to which the note was sent.
+///
+/// Implements the procedure specified in ZIP 307.
+pub fn try_sapling_compact_note_decryption(
+    ivk: &Fs,
+    epk: &edwards::Point<Bls12, PrimeOrder>,
+    cmu: &Fr,
+    enc_ciphertext: &[u8],
+) -> Option<(Note<Bls12>, PaymentAddress<Bls12>)> {
+    let shared_secret = sapling_ka_agree(&ivk, &epk);
+    let key = kdf_sapling(&shared_secret, &epk);
 
-    if note.cm(&JUBJUB) != *cmu {
-        // Published commitment doesn't match calculated commitment
-        return None;
+    let nonce = [0u8; 12];
+    let mut chacha20 = ChaCha20::new(key.as_bytes(), &nonce);
+    // Skip over Poly1305 keying output
+    chacha20.next();
+
+    let mut plaintext = Vec::with_capacity(52);
+    plaintext.extend_from_slice(&enc_ciphertext[0..52]);
+    let keystream = chacha20.next();
+    for i in 0..52 {
+        plaintext[i] ^= keystream.as_bytes()[i];
```

Comment:
Nit: Does it make sense to call `.as_bytes()` outside of the loop? I don't know the details of this function, but I feel like the processor might be doing a little more work than is necessary here.
Edit: I see this changes later so never mind.